rosbag2_cpp/src/rosbag2_cpp/reindexers/sequential_reindexer.cpp

```diff
@@ -0,0 +1,346 @@
+  const rcpputils::fs::path & first_path, const rcpputils::fs::path & second_path)
+{
+  std::regex regex_rule(".*_(\\d+)\\.db3", std::regex_constants::ECMAScript);
+
+  std::smatch first_match;
+  std::smatch second_match;
+
+  auto first_path_string = first_path.string();
+  auto second_path_string = second_path.string();
+
+  auto first_regex_good = std::regex_match(first_path_string, first_match, regex_rule);
+  auto second_regex_good = std::regex_match(second_path_string, second_match, regex_rule);
+
+  // Make sure the paths have regex matches
+  if (!first_regex_good || !second_regex_good) {
+    throw std::runtime_error("Malformed relative file name. Expected numerical identifier.");
+  }
+
+  // Convert database numbers to uint
+  u_int32_t first_db_num = std::stoul(first_match.str(1), nullptr, 10);
+  u_int32_t second_db_num = std::stoul(second_match.str(1), nullptr, 10);
+
+  return first_db_num < second_db_num;
+}
+
+std::vector<rcpputils::fs::path> SequentialReindexer::get_database_files(
+  const rcpputils::fs::path & base_folder)
+{
+  // Look in the uri directory to see what database files are there
+  std::vector<rcpputils::fs::path> output;
+
+  #ifdef WIN32
+  {
+    // Code placed in scope so variables can't accidentally be used elsewhere
+    WIN32_FIND_DATA FindFileData;
+    HANDLE hFind;
+
+    std::string search_dir = base_folder.string() + "\\*";
+    hFind = FindFirstFile(search_dir, &FindFileData);
+    // If an error occurs, we want to abort
+    if (hFind == INVALID_HANDLE_VALUE) {
+      DWORD dwError = GetLastError();
+      std::error_code ec(dwError, std::system_category());
+      throw(std::system_error(ec));
+    }
+
+    // Loop through the directory, collecting file names as we go
+    do {
+      if (ffd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) {
+        // I guess it's a directory in the bag file?
+        // Still, not interested in it.
+        continue;
+      } else {
+        auto temp_path = rcpputils::fs::path(ffd.cFileName);
+
+        // We are ONLY interested in database files
+        if (temp_path.extension().string() != ".db3") {
+          continue;
+        }
+        output.emplace_back(temp_path);
+      }
+    } while (FindNextFile(hFind, &ffd) != 0);
+
+    FindClose(hFind);
+  }
+  #else
+  {
+    // Code placed in scope so variables can't accidentally be used elsewhere
+    auto dirp = opendir(base_folder.string().c_str());
+    // If an error occurs, we want to abort
+    if (dirp == NULL) {
+      throw std::system_error(errno, std::generic_category());
+    }
+    dirent * dp;
+    while ((dp = readdir(dirp)) != NULL) {
+      auto non_const_folder = rcpputils::fs::path(base_folder);
+      auto temp_path = non_const_folder /= rcpputils::fs::path(dp->d_name);
+
+      // We are ONLY interested in database files
+      if (temp_path.extension().string() != ".db3") {
+        continue;
+      }
+
+      output.emplace_back(temp_path);
+    }
+    closedir(dirp);
+  }
+  #endif
```

Comment:
We've recently added code which supersedes this: https://github.com/ros2/rcutils/pull/323